SHARD-2420 : Fixes archiver issue where successful receipts get overwritten by newer failure receipts

abdulazeem-tk4vr opened the pull request:
### **User description**
Problem:
The StoreReceiptData logic on the archiver side is overwriting older receipts stored in the database. This means that if a transaction was previously executed successfully and its receipt was saved, a later failure receipt can replace the original. This may happen due to influencer mode or dirty transaction state from the validator. Although the failure receipt has consensus, its contents may incorrectly indicate that funds were not transferred when they actually were.

Solution:
By checking the transaction status in the receipt, we can determine whether a receipt should be overwritten. This will prevent successful transaction receipts from being replaced by failure receipts.


___

### **PR Type**
Enhancement, Bug fix


___

### **Description**
- Prevents overwriting successful receipts with failure receipts.

- Adds config flag `enableDuplicateReceiptsCheck` for overwrite protection.

- Introduces logging for duplicate receipt overwrite attempts.

- Refactors and improves error handling in receipt storage logic.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Config.ts</strong><dd><code>Add config flag for duplicate receipt overwrite protection</code></dd></summary>
<hr>

src/Config.ts

<li>Adds <code>enableDuplicateReceiptsCheck</code> config flag to control duplicate <br>receipt logic.<br> <li> Sets default value for the new config flag.


</details>


  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/242/files#diff-b83e1482492860acb73eb6a72535d0dd31b4fc971525ff13c35fb97222172d39">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Collector.ts</strong><dd><code>Add receipt overwrite protection and duplicate logging</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Data/Collector.ts

<li>Implements <code>checkIfValidOverwrite</code> to prevent overwriting success <br>receipts with failures.<br> <li> Integrates overwrite check into <code>storeReceiptData</code> using the new config <br>flag.<br> <li> Adds logging for duplicate receipt scenarios.<br> <li> Refactors receipt storage logic for clarity and error handling.


</details>


  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/242/files#diff-186c711dbe48d8ffd4c6488e2d29090dae64abd946a5cf185a549af2c556a72e">+82/-30</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>DataLogWriter.ts</strong><dd><code>Add log writer for receipt overwrite attempts</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Data/DataLogWriter.ts

<li>Adds <code>ReceiptOverwriteLogWriter</code> for logging overwrite attempts.<br> <li> Initializes new log writer in <code>initDataLogWriter</code>.


</details>


  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/242/files#diff-ef69489a30801b1daea305d716a10ba13964ef63c7cd3d6179ef6f38e09eb107">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

abdulazeem-tk4vr committed:
feat: add receipt overwrite protection

Implements a mechanism to prevent overwriting valid receipts with potentially invalid ones, improving data integrity.
Introduces a check to ensure that an existing receipt with a success status (1) cannot be overwritten by a new receipt, even if the new receipt has a failure status (0).
Adds logging for duplicate receipts to aid in debugging and monitoring.

abdulazeem-tk4vr committed:
feat: improve error handling and have it config enabled

abdulazeem-tk4vr committed:
feat: renames config flag for duplicate receipt check

Renames the `enableStatusLogic` configuration flag to
`enableDuplicateReceiptsCheck` to more accurately reflect its
purpose: enabling or disabling the duplicate receipt check in
`storeReceiptData` to prevent overwriting successful receipts
with failures.

justin-shardeum commented on src/Data/Collector.ts:
```diff
@@ -648,6 +654,49 @@ export const verifyArchiverReceipt = async (
+  } catch (error) {
+    // Log the error with context about what operation was attempted
+    Logger.mainLogger.error(`Error in checkIfValidOverwrite for txId ${txId}: ${error.message}`, error)
+    if (nestedCountersInstance) nestedCountersInstance.countEvent('receipt', 'checkIfValidOverwrite_error')
```
This would be a good place to add an alert. But no need to do it in this commit.

abdulazeem-tk4vr committed:
feat: improves data logging and duplicate receipt handling.

Enhances the reliability of data logging by adding a try-catch block to the `writeToLog` function to handle potential errors during the writing process.

Provides additional logging for duplicate receipts, including the content of both the existing and incoming receipts, to aid in debugging and analysis. This improves visibility into potential data inconsistencies and helps to identify the root cause of duplicate receipts.
